File: android/app/src/main/kotlin/com/goalmaker/app/ui/composer/ComposerChips.kt
package com.goalmaker.app.ui.composer

import androidx.compose.runtime.Composable
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.res.stringResource
import com.goalmaker.app.R
import com.goalmaker.app.application.planning.AreaItem
import com.goalmaker.app.domain.composer.ComposerDraft
import com.goalmaker.app.domain.composer.ComposerSpan
import com.goalmaker.app.domain.composer.SpanKind
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import java.time.format.TextStyle
import java.util.Locale

/** The preview chips for [draft], in the order the item reads: when, how often, where, tags, flags. */
@Composable
fun composerChips(line: String, draft: ComposerDraft, today: LocalDate, areas: List<AreaItem>, tagNames: List<String>): List<ComposerChip> {
    val locale = LocalConfiguration.current.locales[0]
    val spans = draft.spans
    fun of(kind: SpanKind) = spans.filter { it.kind == kind }
    val newNote = stringResource(R.string.composer_new)
    val laterNote = stringResource(R.string.composer_later)
    val chips = mutableListOf<ComposerChip>()

    draft.command?.let { command ->
        val note = stringResource(if (command.known) R.string.composer_soon else R.string.composer_unknown_command)
        chips += ComposerChip(SpanKind.COMMAND, "/${command.name}", note, muted = true, areaColorId = null, spans = of(SpanKind.COMMAND))
        return chips
    }
    draft.plannedDate?.let { date ->
        val label = when (date) {
            today -> stringResource(R.string.composer_date_today)
            today.plusDays(1) -> stringResource(R.string.composer_date_tomorrow)
            else -> DateTimeFormatter.ofPattern(if (date.year == today.year) "EEE d MMM" else "EEE d MMM yyyy", locale).format(date)
        }
        chips += ComposerChip(SpanKind.DATE, label, null, false, null, of(SpanKind.DATE))
    }
    draft.plannedTime?.let { time ->
        chips += ComposerChip(SpanKind.TIME, DateTimeFormatter.ofLocalizedTime(FormatStyle.SHORT).withLocale(locale).format(time), null, false, null, of(SpanKind.TIME))
    }
    draft.repeat?.let { rule ->
        chips += ComposerChip(SpanKind.REPEAT, describeRepeat(rule, locale), null, false, null, of(SpanKind.REPEAT))
    }
    draft.area?.let { name ->
        val existing = areas.firstOrNull { it.name.trim().lowercase(Locale.ROOT) == name.trim().lowercase(Locale.ROOT) }
        chips += ComposerChip(SpanKind.AREA, existing?.name ?: name, if (existing == null) newNote else null, false, existing?.colorId, of(SpanKind.AREA))
    }
    val known = tagNames.map { it.lowercase(Locale.ROOT) }.toSet()
    draft.tags.forEach { tag ->
        val tagSpans = of(SpanKind.TAG).filter { span ->
            line.substring(span.start + 1, span.end).trimEnd { it in ".,;:!?" }.lowercase(Locale.ROOT) == tag.lowercase(Locale.ROOT)
        }
        chips += ComposerChip(SpanKind.TAG, "#$tag", if (tag.lowercase(Locale.ROOT) in known) null else newNote, false, null, tagSpans)
    }
    if (draft.topPriority) {
        chips += ComposerChip(SpanKind.PRIORITY, stringResource(R.string.composer_priority), null, false, null, of(SpanKind.PRIORITY))
    }
    draft.project?.let { name ->
        chips += ComposerChip(SpanKind.PROJECT, "+$name", laterNote, true, null, of(SpanKind.PROJECT))
    }
    if (draft.idea) {
        chips += ComposerChip(SpanKind.IDEA, stringResource(R.string.composer_idea), laterNote, true, null, of(SpanKind.IDEA))
    }
    return chips
}

/** The line without the given parts, spaces tidied. */
fun removeParts(line: String, spans: List<ComposerSpan>): String {
    val text = StringBuilder(line)
    spans.sortedByDescending { it.start }.forEach { text.delete(it.start, it.end) }
    return text.toString().replace(WHITESPACE, " ").trim()
}

private val WHITESPACE = Regex("\\s+")

@Composable
private fun describeRepeat(rule: String, locale: Locale): String {
    val parts = rule.split(';').associate { it.substringBefore('=') to it.substringAfter('=') }
    val interval = parts["INTERVAL"]?.toIntOrNull() ?: 1
    return when (parts["FREQ"]) {
        "DAILY" -> if (interval > 1) stringResource(R.string.repeat_every_n_days, interval) else stringResource(R.string.repeat_daily)
        "WEEKLY" -> {
            val codes = parts["BYDAY"].orEmpty().split(',')
            val days = codes.mapNotNull { code -> DayOfWeek.entries.firstOrNull { it.name.startsWith(code) } }
            val names = days.joinToString(", ") { it.getDisplayName(TextStyle.SHORT, locale) }
            when {
                interval > 1 -> stringResource(R.string.repeat_every_n_weeks_on, interval, names)
                days.toSet() == WORK_WEEK -> stringResource(R.string.repeat_weekdays)
                days.toSet() == WEEKEND -> stringResource(R.string.repeat_weekend)
                days.size == 1 -> stringResource(R.string.repeat_every_day_name, days.single().getDisplayName(TextStyle.FULL, locale))
                else -> stringResource(R.string.repeat_every_day_name, names)
            }
        }
        "MONTHLY" -> {
            val day = parts["BYMONTHDAY"]?.toIntOrNull() ?: 1
            if (interval > 1) stringResource(R.string.repeat_every_n_months_on, interval, day) else stringResource(R.string.repeat_monthly_on, day)
        }
        else -> rule
    }
}

private val WORK_WEEK = setOf(DayOfWeek.MONDAY, DayOfWeek.TUESDAY, DayOfWeek.WEDNESDAY, DayOfWeek.THURSDAY, DayOfWeek.FRIDAY)
private val WEEKEND = setOf(DayOfWeek.SATURDAY, DayOfWeek.SUNDAY)

File: windows/src/GoalMaker.App/ViewModels/ComposerChips.cs
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;
using System.Windows.Media;
using GoalMaker.App.Localization;
using GoalMaker.Core.Composer;
using GoalMaker.Core.Planning;
using Wpf.Ui.Controls;

namespace GoalMaker.App.ViewModels;

/// <summary>Builds the composer's preview chips (docs/composer.md), in the order the item reads.</summary>
public static partial class ComposerChips
{
    private static readonly DayOfWeek[] MondayFirst =
        [DayOfWeek.Monday, DayOfWeek.Tuesday, DayOfWeek.Wednesday, DayOfWeek.Thursday, DayOfWeek.Friday, DayOfWeek.Saturday, DayOfWeek.Sunday];

    public static IReadOnlyList<ComposerChipViewModel> Build(
        string line,
        ComposerDraft draft,
        DateOnly today,
        IReadOnlyList<AreaItem> areas,
        IReadOnlyList<string> tagNames,
        IStrings strings,
        Func<string, Brush?> areaBrush,
        Action<ComposerChipViewModel> remove)
    {
        var chips = new List<ComposerChipViewModel>();
        IReadOnlyList<ComposerSpan> Of(SpanKind kind) => [.. draft.Spans.Where(span => span.Kind == kind)];
        void Add(SpanKind kind, string label, string? note, SymbolRegular symbol, IReadOnlyList<ComposerSpan> spans, Brush? brush = null, bool muted = false) =>
            chips.Add(new ComposerChipViewModel(
                kind,
                note is null ? label : $"{label} · {note}",
                strings.Get("Composer.Remove", label),
                symbol,
                brush,
                muted,
                spans,
                remove));

        if (draft.Command is { } command)
        {
            var runs = command.Name == PlanRules.Command;
            var note = strings.Get(runs ? "Composer.OpensPlan" : command.Known ? "Composer.Soon" : "Composer.UnknownCommand");
            Add(SpanKind.Command, "/" + command.Name, note, runs ? SymbolRegular.CalendarEdit24 : SymbolRegular.Code24, Of(SpanKind.Command), muted: !runs);
            return chips;
        }

        if (draft.PlannedDate is { } date)
        {
            var label = date == today ? strings.Get("Composer.Today")
                : date == today.AddDays(1) ? strings.Get("Composer.Tomorrow")
                : date.ToString(date.Year == today.Year ? "ddd d MMM" : "ddd d MMM yyyy", CultureInfo.CurrentCulture);
            Add(SpanKind.Date, label, null, SymbolRegular.CalendarLtr24, Of(SpanKind.Date));
        }

        if (draft.PlannedTime is { } time)
        {
            Add(SpanKind.Time, time.ToString("t", CultureInfo.CurrentCulture), null, SymbolRegular.Clock24, Of(SpanKind.Time));
        }

        if (draft.Repeat is { } rule)
        {
            Add(SpanKind.Repeat, DescribeRepeat(rule, strings), null, SymbolRegular.ArrowRepeatAll24, Of(SpanKind.Repeat));
        }

        if (draft.Area is { } areaName)
        {
            var existing = areas.FirstOrDefault(area => Key(area.Name) == Key(areaName));
            Add(SpanKind.Area, existing?.Name ?? areaName, existing is null ? strings.Get("Composer.New") : null, SymbolRegular.Circle24, Of(SpanKind.Area),
                existing is null ? null : areaBrush(existing.ColorId));
        }

        var known = tagNames.Select(Key).ToHashSet(StringComparer.Ordinal);
        foreach (var tag in draft.Tags)
        {
            IReadOnlyList<ComposerSpan> spans = [.. Of(SpanKind.Tag).Where(span => Key(line[(span.Start + 1)..span.End].TrimEnd('.', ',', ';', ':', '!', '?')) == Key(tag))];
            Add(SpanKind.Tag, "#" + tag, known.Contains(Key(tag)) ? null : strings.Get("Composer.New"), SymbolRegular.NumberSymbol24, spans);
        }

        if (draft.TopPriority)
        {
            Add(SpanKind.Priority, strings.Get("Composer.Priority"), null, SymbolRegular.Flag24, Of(SpanKind.Priority));
        }

        if (draft.Project is { } project)
        {
            Add(SpanKind.Project, "+" + project, strings.Get("Composer.Later"), SymbolRegular.Folder24, Of(SpanKind.Project), muted: true);
        }

        if (draft.Idea)
        {
            Add(SpanKind.Idea, strings.Get("Composer.Idea"), strings.Get("Composer.Later"), SymbolRegular.Lightbulb24, Of(SpanKind.Idea), muted: true);
        }

        return chips;
    }

    /// <summary>The line without the given parts, spaces tidied.</summary>
    public static string RemoveParts(string line, IEnumerable<ComposerSpan> spans)
    {
        var text = new StringBuilder(line);
        foreach (var span in spans.OrderByDescending(span => span.Start))
        {
            text.Remove(span.Start, span.End - span.Start);
        }

        return Whitespace().Replace(text.ToString(), " ").Trim();
    }

    private static string Key(string name) => name.Trim().ToLowerInvariant();

    /// <summary>A repeat rule in words, for the composer's chip and the task's detail page.</summary>
    internal static string DescribeRepeat(string rule, IStrings strings)
    {
        var parts = rule.Split(';').Select(part => part.Split('=', 2)).ToDictionary(pair => pair[0], pair => pair.Length > 1 ? pair[1] : string.Empty, StringComparer.Ordinal);
        var interval = parts.TryGetValue("INTERVAL", out var text) && int.TryParse(text, CultureInfo.InvariantCulture, out var n) ? n : 1;
        var format = CultureInfo.CurrentCulture.DateTimeFormat;
        switch (parts.GetValueOrDefault("FREQ"))
        {
            case "DAILY":
                return interval > 1 ? strings.Get("Repeat.EveryNDays", interval) : strings.Get("Repeat.Daily");
            case "WEEKLY":
                var codes = parts.GetValueOrDefault("BYDAY", string.Empty).Split(',');
                var days = MondayFirst.Where(day => codes.Contains(day.ToString()[..2].ToUpperInvariant())).ToList();
                var names = string.Join(", ", days.Select(format.GetAbbreviatedDayName));
                if (interval > 1)
                {
                    return strings.Get("Repeat.EveryNWeeksOn", interval, names);
                }

                if (days.Count == 5 && days.All(day => day is not (DayOfWeek.Saturday or DayOfWeek.Sunday)))
                {
                    return strings.Get("Repeat.Weekdays");
                }

                if (days.Count == 2 && days.All(day => day is DayOfWeek.Saturday or DayOfWeek.Sunday))
                {
                    return strings.Get("Repeat.Weekend");
                }

                return strings.Get("Repeat.EveryDayName", days.Count == 1 ? format.GetDayName(days[0]) : names);
            case "MONTHLY":
                var monthDay = parts.TryGetValue("BYMONTHDAY", out var dayText) && int.TryParse(dayText, CultureInfo.InvariantCulture, out var d) ? d : 1;
                return interval > 1 ? strings.Get("Repeat.EveryNMonthsOn", interval, monthDay) : strings.Get("Repeat.MonthlyOn", monthDay);
            default:
                return rule;
        }
    }

    [GeneratedRegex(@"\s+", RegexOptions.CultureInvariant)]
    private static partial Regex Whitespace();
}
